File: tests/regenerate/gen_erf_normal_kernels.c
/*
 * Reference tables for the erf and standard-normal kernels:
 * error_f, error_fc, error_fc_scaled, cumnor, dinvnr.
 *
 * Build and run via tests/regenerate/regenerate.sh.
 */

#include "refs/cdflib.h"
#include <stdio.h>
#include <stdlib.h>

static FILE* open_csv(const char* relpath, const char* header) {
    FILE* f = fopen(relpath, "w");
    if (!f) {
        fprintf(stderr, "could not open %s for writing\n", relpath);
        exit(1);
    }
    fprintf(f, "# %s\n", header);
    return f;
}

static void write_row3(FILE* f, double a, double b, double c) {
    /* %.17g is the round-trip-safe printf for f64 — every distinct f64
       value gets a distinct decimal representation. */
    fprintf(f, "%.17g,%.17g,%.17g\n", a, b, c);
}

static void write_row2(FILE* f, double a, double b) {
    fprintf(f, "%.17g,%.17g\n", a, b);
}

int main(void) {
    /* ------------------------------------------------------------ error_f */
    FILE* erff = open_csv("tests/data/error_f.csv", "x, erf(x)");
    /* Dense grid spanning all three branches (|x| ≤ 0.5, ≤ 4, < 5.8, ≥ 5.8)
       plus the saturation cutoff. Step 0.0625 to land on f64-exact values. */
    for (double x = -6.0; x <= 6.0 + 1e-12; x += 0.0625) {
        double xx = x;
        write_row2(erff, x, error_f(&xx));
    }
    fclose(erff);

    /* ----------------------------------------------------------- error_fc */
    FILE* erfcf = open_csv("tests/data/error_fc.csv", "x, erfc(x)");
    int ind0 = 0;
    /* erfc needs more extreme positive coverage because the right tail is
       what we actually care about for accuracy. */
    for (double x = -6.0; x <= 30.0 + 1e-12; x += 0.0625) {
        double xx = x;
        write_row2(erfcf, x, error_fc(&ind0, &xx));
    }
    fclose(erfcf);

    /* ----------------------------------------------------- error_fc_scaled */
    FILE* erfcs = open_csv("tests/data/error_fc_scaled.csv", "x, erfc(x)*exp(x^2)");
    int ind1 = 1;
    /* Scaled form is well-defined for arbitrarily large positive x; large
       negative x explodes via 2*exp(x^2), so cap the negative grid. */
    for (double x = -6.0; x <= 60.0 + 1e-12; x += 0.0625) {
        double xx = x;
        write_row2(erfcs, x, error_fc(&ind1, &xx));
    }
    fclose(erfcs);

    /* ------------------------------------------------------------- cumnor */
    FILE* cnf = open_csv("tests/data/cumnor.csv", "x, cum, ccum");
    /* Cover all three regions of cumnor (|x| ≤ 0.66291, ≤ √32, > √32) plus
       deep tails up to |x| = 38 where the result is near machine min. */
    for (double x = -38.0; x <= 38.0 + 1e-12; x += 0.0625) {
        double cum, ccum;
        double xx = x;
        cumnor(&xx, &cum, &ccum);
        write_row3(cnf, x, cum, ccum);
    }
    fclose(cnf);

    /* ------------------------------------------------------------- dinvnr */
    /* Parametrize by x rather than p, then store (p, q, x). This keeps the
       small tail represented honestly: at x = 8 we have q ≈ 6e-16, which
       is unrepresentable as `1 - p` because p is rounded to 1.0. */
    FILE* dnf = open_csv("tests/data/dinvnr.csv", "p, q, x");
    for (double x = -7.0; x <= 7.0 + 1e-12; x += 0.0625) {
        double cum, ccum;
        double xx = x;
        cumnor(&xx, &cum, &ccum);
        /* Skip rows where cum or ccum has saturated to exactly 0 or 1,
           since they have no information for the inverse. */
        if (cum <= 0.0 || cum >= 1.0 || ccum <= 0.0 || ccum >= 1.0) continue;
        double back = dinvnr(&cum, &ccum);
        write_row3(dnf, cum, ccum, back);
    }
    fclose(dnf);

    fprintf(stderr, "wrote 5 tables under tests/data/\n");
    return 0;
}

File: tests/regenerate/gen_erf_normal_kernels.f90
! Reference tables for the erf and standard-normal kernels:
! error_f, error_fc, error_fc_scaled, cumnor, dinvnr.
!
! Build and run via tests/regenerate/regenerate.sh.

program gen_erf_normal_kernels
  implicit none
  integer, parameter :: rk = kind(1.0d0)

  real(kind=rk), external :: error_f, error_fc, dinvnr
  external :: cumnor

  call gen_error_f()
  call gen_error_fc()
  call gen_error_fc_scaled()
  call gen_cumnor()
  call gen_dinvnr()

  write(0, '(a)') 'wrote 5 tables under tests/data/'

contains

  ! Write a single double-precision value to `unit` followed by a comma,
  ! or by a newline if `last` is true. Uses 17 significant digits so each
  ! distinct f64 round-trips through decimal.
  subroutine putval(unit, v, last)
    integer, intent(in) :: unit
    real(kind=rk), intent(in) :: v
    logical, intent(in) :: last
    character(len=32) :: buf
    ! es25.17e3 = sign + digit + '.' + 17 frac + 'E' + signed-3 = 25 chars
    write(buf, '(es25.17e3)') v
    if (last) then
      write(unit, '(a)') trim(adjustl(buf))
    else
      write(unit, '(a, a)', advance='no') trim(adjustl(buf)), ','
    end if
  end subroutine putval

  subroutine gen_error_f()
    integer :: unit
    real(kind=rk) :: x
    open(newunit=unit, file='tests/data/error_f.csv', status='replace', action='write')
    write(unit, '(a)') '# x, erf(x)'
    x = -6.0_rk
    do while (x <= 6.0_rk + 1.0e-12_rk)
      call putval(unit, x, .false.)
      call putval(unit, error_f(x), .true.)
      x = x + 0.0625_rk
    end do
    close(unit)
  end subroutine gen_error_f

  subroutine gen_error_fc()
    integer :: unit, ind0
    real(kind=rk) :: x
    ind0 = 0
    open(newunit=unit, file='tests/data/error_fc.csv', status='replace', action='write')
    write(unit, '(a)') '# x, erfc(x)'
    x = -6.0_rk
    do while (x <= 30.0_rk + 1.0e-12_rk)
      call putval(unit, x, .false.)
      call putval(unit, error_fc(ind0, x), .true.)
      x = x + 0.0625_rk
    end do
    close(unit)
  end subroutine gen_error_fc

  subroutine gen_error_fc_scaled()
    integer :: unit, ind1
    real(kind=rk) :: x
    ind1 = 1
    open(newunit=unit, file='tests/data/error_fc_scaled.csv', status='replace', action='write')
    write(unit, '(a)') '# x, erfc(x)*exp(x^2)'
    x = -6.0_rk
    do while (x <= 60.0_rk + 1.0e-12_rk)
      call putval(unit, x, .false.)
      call putval(unit, error_fc(ind1, x), .true.)
      x = x + 0.0625_rk
    end do
    close(unit)
  end subroutine gen_error_fc_scaled

  subroutine gen_cumnor()
    integer :: unit
    real(kind=rk) :: x, cum, ccum
    open(newunit=unit, file='tests/data/cumnor.csv', status='replace', action='write')
    write(unit, '(a)') '# x, cum, ccum'
    x = -38.0_rk
    do while (x <= 38.0_rk + 1.0e-12_rk)
      call cumnor(x, cum, ccum)
      call putval(unit, x, .false.)
      call putval(unit, cum, .false.)
      call putval(unit, ccum, .true.)
      x = x + 0.0625_rk
    end do
    close(unit)
  end subroutine gen_cumnor

  subroutine gen_dinvnr()
    ! Parametrize by x rather than p; store (cum, ccum, back). Skips rows
    ! where cum/ccum has saturated to 0 or 1.
    integer :: unit
    real(kind=rk) :: x, cum, ccum, back
    open(newunit=unit, file='tests/data/dinvnr.csv', status='replace', action='write')
    write(unit, '(a)') '# p, q, x'
    x = -7.0_rk
    do while (x <= 7.0_rk + 1.0e-12_rk)
      call cumnor(x, cum, ccum)
      if (cum > 0.0_rk .and. cum < 1.0_rk .and. ccum > 0.0_rk .and. ccum < 1.0_rk) then
        back = dinvnr(cum, ccum)
        call putval(unit, cum, .false.)
        call putval(unit, ccum, .false.)
        call putval(unit, back, .true.)
      end if
      x = x + 0.0625_rk
    end do
    close(unit)
  end subroutine gen_dinvnr

end program gen_erf_normal_kernels
